stackslib/src/chainstate/stacks/boot/pox_4_tests.rs

```diff
@@ -1234,3 +1249,405 @@ fn balances_from_keys(
+        vec![
+            Value::UInt(min_ustx),
+            pox_addr,
+            Value::UInt(block_height as u128),
+            Value::UInt(2),
+            signer_key_val.clone(),
+        ],
+    )];
+
+    let latest_block = peer.tenure_with_txs(&txs, &mut coinbase_nonce);
+    let stacking_state = get_stacking_state_pox_4(
+        &mut peer,
+        &latest_block,
+        &key_to_stacks_addr(stacker_key).to_account_principal(),
+    )
+    .expect("No stacking state, stack-stx failed")
+    .expect_tuple();
+
+    let state_signer_key = stacking_state.get("signer-key").unwrap();
+    assert_eq!(state_signer_key.to_string(), signer_key_val.to_string());
+}
+
+#[test]
+fn delegate_stx_signer_key() {
+    let lock_period = 2;
+    let (burnchain, mut peer, keys, latest_block, block_height, mut coinbase_nonce) =
+        prepare_pox4_test(function_name!());
+
+    let stacker_nonce = 0;
+    let stacker_key = &keys[0];
+    let delegate_key = &keys[1];
+    let delegate_principal = PrincipalData::from(key_to_stacks_addr(delegate_key));
+
+    // (define-public (delegate-stx (amount-ustx uint)
+    //                          (delegate-to principal)
+    //                          (until-burn-ht (optional uint))
+    //                          (pox-addr (optional { version: (buff 1), hashbytes: (buff 32) }))
+    //                          (signer-key (optional (buff 33))))
+    let pox_addr = make_pox_addr(
+        AddressHashMode::SerializeP2WSH,
+        key_to_stacks_addr(delegate_key).bytes,
+    );
+    let signer_key_val = Value::buff_from(vec![
+        0x03, 0xa0, 0xf9, 0x81, 0x8e, 0xa8, 0xc1, 0x4a, 0x82, 0x7b, 0xb1, 0x44, 0xae, 0xc9, 0xcf,
+        0xba, 0xeb, 0xa2, 0x25, 0xaf, 0x22, 0xbe, 0x18, 0xed, 0x78, 0xa2, 0xf2, 0x98, 0x10, 0x6f,
+        0x4e, 0x28, 0x1b,
+    ])
+    .unwrap();
+    let txs = vec![make_pox_4_contract_call(
+        stacker_key,
+        stacker_nonce,
+        "delegate-stx",
+        vec![
+            Value::UInt(100),
+            delegate_principal.clone().into(),
+            Value::none(),
+            Value::Optional(OptionalData {
+                data: Some(Box::new(pox_addr)),
+            }),
+            Value::Optional(OptionalData {
+                data: Some(signer_key_val.clone().into()),
+            }),
+        ],
+    )];
+
+    let latest_block = peer.tenure_with_txs(&txs, &mut coinbase_nonce);
+    let delegation_state = get_delegation_state_pox_4(
+        &mut peer,
+        &latest_block,
+        &key_to_stacks_addr(stacker_key).to_account_principal(),
+    )
+    .expect("No delegation state, delegate-stx failed")
+    .expect_tuple();
+
+    let state_signer_key_optional = delegation_state.get("signer-key").unwrap();
+    assert_eq!(
+        state_signer_key_optional.to_string(),
+        Value::Optional(OptionalData {
+            data: Some(Box::new(signer_key_val))
+        })
+        .to_string()
+    );
+}
+
+#[test]
+fn stack_extend_signer_key() {
```

Comment:
This is a great initiative, @MarvinJanssen 👍! It aligns well with the goals for #4092 too.

To further enhance test readability and maintainability, I suggest we consider creating some common abstractions, maybe even a small DSL, in future PRs. /cc @jbencin @setzeus 

Here's an example of how we could refactor a test:

```rust
// Refactored test
#[test]
fn stack_stx_signer_key() {
    // Arrange
    let mut env = setup_test_environment("stack_stx_signer_key");
    let (signer_key_val, txs) = prepare_stack_stx_data(&mut env);
    
    // Act
    let latest_block = execute_transactions(&mut env, &txs);
    let stacking_state = env
        .get_stacking_state(&latest_block, &env.stacker_key()
        .to_account_principal())
        .expect("No stacking state").expect_tuple();

    // Assert
    assert_signer_key(
        &stacking_state.get("signer-key").unwrap(),
        &signer_key_val);
}

// Where:

// Common setup
fn setup_test_environment(test_name: &str) -> TestEnvironment {
    // Setup logic here
}

// Specific data preparation
fn prepare_stack_stx_data(env: &mut TestEnvironment) -> (Value, Vec<Transaction>) {
    // Data preparation logic
}

// Execute actions
fn execute_transactions(env: &mut TestEnvironment, txs: &[Transaction]) -> LatestBlock {
    // Transaction execution logic
}

// Assertion functions
fn assert_signer_key(expected: &Value, actual: &Value) {
    // Assertion logic
}
```

Reply:
Yes I fully agree. I tried to find a middle ground to keep the PR compact. Also I want to underscore that the vast majority of pox4 unit tests will/are written in Clarity (as you know), most of which has/will be worked on by @setzeus. The tooling to make those tests part of the repo and CI is currently being done per @wileyj's instructions. I would say, once all of that is being brought in and worked on, we will be able to identify which parts cannot be done in Clarity tests (like the StackerDB god-mode stuff) and create common abstractions to service those tests.